### PR TITLE
fix(sw360): wrong function called when updating fixed

### PR DESCRIPTION
**Pull request by neubs-bsi**
Instead of calling `updateRelease`, `addRelease` was called and hence created a 409 Conflict Response when running same project twice. 

Error isn't occuring anymore. There's a new one now. Patch isn't functioning still. But normal getOrCreateRelease is (which is relevant for antenna users)

### CODE REVIEW

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/adapter/SW360ReleaseClientAdapter.java [hunk 1]
```diff
@@ -40,8 +40,10 @@ public SW360Release addRelease(Artifact artifact, SW360Component sw360Component,
     }
 
     public SW360Release updateRelease(SW360Release release, Artifact artifact, SW360Component sw360Component, Set<String> sw360LicenseIds, HttpHeaders header) throws AntennaException {
+        release = release.mapExternalIdsOnVariable();
```
Comment (maxhbr) [inline modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/adapter/SW360ReleaseClientAdapter.java]:
```suggestion
        SW360Release mappedRelease = release.mapExternalIdsOnVariable();
```
to avoid mutation

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/SW360Client.java [hunk 2]
```diff
@@ -65,6 +66,11 @@ public SW360Client(boolean proxyUse, String proxyHost, int proxyPort) {
     }
 
     protected <T> ResponseEntity<T> doRestPATCH(String url, HttpEntity<?> httpEntity, ParameterizedTypeReference<T> responseType) {
-        return doRestCall(url, HttpMethod.PATCH, httpEntity, responseType);
+        RestTemplate restTemplate = new RestTemplate();
+
+        HttpComponentsClientHttpRequestFactory requestFactory = new HttpComponentsClientHttpRequestFactory();
```
Comment (maxhbr) [inline modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/SW360Client.java]: Why not instead change the `SimpleClientHttpRequestFactory requestFactory` the constructor to `HttpComponentsClientHttpRequestFactory requestFactory`?

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java [hunk 3]
```diff
@@ -192,26 +201,69 @@ public void setCopyrights(String copyrights) {
...
-        declaredLicense = Optional.of(releaseWithPrecedence.getDeclaredLicense()).orElse(declaredLicense);
-        observedLicense = Optional.of(releaseWithPrecedence.getObservedLicense()).orElse(observedLicense);
-        releaseTagUrl = Optional.of(releaseWithPrecedence.getReleaseTagUrl()).orElse(releaseTagUrl);
-        softwareHeritageId = Optional.of(releaseWithPrecedence.getSoftwareHeritageId()).orElse(softwareHeritageId);
-        hashes = Optional.of(releaseWithPrecedence.getHashes()).orElse(hashes);
-        clearingState = Optional.of(releaseWithPrecedence.getClearingState()).orElse(clearingState);
-        changeStatus = Optional.of(releaseWithPrecedence.getChangeStatus()).orElse(changeStatus);
-        copyrights = Optional.of(releaseWithPrecedence.getCopyrights()).orElse(copyrights);
-        externalIds = Optional.of(releaseWithPrecedence.getExternalIds()).orElse(externalIds);
+        cpeid = Optional.ofNullable(releaseWithPrecedence.getCpeid()).orElse(Optional.ofNullable(cpeid).orElse(null));
```
Comment (maxhbr) [inline modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java]: every line has the same pattern, maybe extract into a function:
```
    private T fun(SW360Release release, Function<SW360Release,T> getter);
```

Then you would only have to write `cpeid = fun(releaseWithPrecedence, SW360Release::getCpeId)` which is cleaner and more readable

Comment (maxhbr) [reply modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java]: The call to `Optional.ofNullable(cpeid).orElse(null)` should be a no-op and always equal to `cpeid`, i.e.
- if `cpeid` is not null, it just returns `cpeid`
- if `cpeid` is null, it falls in the `orElse` case and returns `null` which is still equal to `cpeid`

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/utils/RestUtils.java [hunk 4]
```diff
@@ -123,6 +124,9 @@
             externalIds.put(SW360Attributes.RELEASE_EXTERNAL_ID_COPYRIGHTS, sw360Release.getCopyrights());
         }
 
+        if(externalIds.isEmpty()) {
+            return null;
```
Comment (maxhbr) [inline modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/utils/RestUtils.java]: why should `null` be better than a empty map? This is just prone to NPEs and requests for more null-checks later

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java [hunk 5]
```diff
@@ -192,26 +201,69 @@ public void setCopyrights(String copyrights) {
...
+                default:
+                    if (key.startsWith(SW360Attributes.RELEASE_EXTERNAL_ID_HASHES)) {
+                        if(hashes == null) {
+                            hashes = Collections.singleton(value);
+                        } else {
+                            hashes.add(value);
+                        }
+                    } else if (SW360CoordinateKeysToArtifactCoordinates.getValues().contains(key)) {
+                        if(coordinates == null) {
+                            coordinates = new HashMap<String, String>() {{
```
Comment (maxhbr) [inline modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java]: I think this way of initializing is deprecated (or at least it is ugly).
I would even prefer:
```suggestion
                            coordinates = new HashMap<String, String>();
                            coordinates.put(key, value);
```

(and even more I would prefer an immutable map, but that is a completely different story ;) )

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/SW360Client.java (hunk 2 above)
Comment (neubs-bsi) [reply modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/SW360Client.java]: `HttpComponentsClientHttpRequestFactory` doesn't have a possibility to set a proxy

Comment (maxhbr) [reply modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/SW360Client.java]: This means that this fix only works without a proxy?

Comment (maxhbr) [reply modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/SW360Client.java]: Maybe throw an unsupported operation exception, if patch is requested with a proxy

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java (hunk 3 above)
Comment (neubs-bsi) [reply modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java]: when I put `orElse(cpeid)` in there, it caused a null pointer excpetion when it wasn't set. Hence I changed it to this mess, that worked.

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java [hunk 6]
```diff
@@ -201,30 +210,77 @@ public void setCopyrights(String copyrights) {
...
+        softwareHeritageId = fun(releaseWithPrecedence, SW360Release::getSoftwareHeritageId);
+        hashes = fun(releaseWithPrecedence, SW360Release::getHashes);
+        clearingState = fun(releaseWithPrecedence, SW360Release::getClearingState);
+        changeStatus = fun(releaseWithPrecedence, SW360Release::getChangeStatus);
+        copyrights =  fun(releaseWithPrecedence, SW360Release::getCopyrights);
 
         return this;
     }
 
+    private <T> T fun(SW360Release release, Function<SW360Release, T> getter) {
```
Comment (maxhbr) [inline modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java]: maybe find a better name

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java [hunk 7]
```diff
@@ -201,30 +210,77 @@ public void setCopyrights(String copyrights) {
...
+        return Optional.ofNullable(getter.apply(release)).orElse(Optional.ofNullable(getter.apply(this)).orElse(null));
```
Comment (maxhbr) [inline modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java]: better formatting
```suggestion
        return Optional.ofNullable(getter.apply(release))
            .orElse(Optional.ofNullable(getter.apply(this))
            .orElse(null));
```

And should still be equal to:
```suggestion
        return Optional.ofNullable(getter.apply(release))
            .orElse(getter.apply(this));
```

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/utils/RestUtils.java [hunk 8]
```diff
@@ -123,6 +124,9 @@
             externalIds.put(SW360Attributes.RELEASE_EXTERNAL_ID_COPYRIGHTS, sw360Release.getCopyrights());
         }
 
+        if(externalIds.isEmpty()) {
+            return Collections.emptyMap();
```
Comment (maxhbr) [inline modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/utils/RestUtils.java]: this only returns an immutable map instead off the empty `externalIds` map? Why is this here?

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java [hunk 9]
```diff
@@ -201,30 +210,77 @@ public void setCopyrights(String copyrights) {
...
 
     public boolean shareIdentifier(SW360Release releaseCompare) {
-        return this.componentId.equals(releaseCompare.getComponentId())
-                && this.name.equals(releaseCompare.getName())
-                && this.version.equals(releaseCompare.getVersion());
+        return this.name.equals(Optional.of(releaseCompare.getName()).orElse(""))
+                && this.version.equals(Optional.of(releaseCompare.getVersion()).orElse(""));
+    }
+
+    public SW360Release mapExternalIdsOnVariable() {
```
Comment (maxhbr) [inline modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java]: Why not always call this method after `setExternalIds()` is called?

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java [hunk 10]
```diff
@@ -201,30 +210,77 @@ public void setCopyrights(String copyrights) {
...
+                            hashes.add(value);
+                        }
+                    } else if (SW360CoordinateKeysToArtifactCoordinates.getValues().contains(key)) {
+                        if(coordinates == null) {
+                            coordinates = new HashMap<String, String>();
+                            coordinates.put(key, value);
+                        } else {
+                            coordinates.put(key, value);
+                        }
+                        coordinates = Collections.unmodifiableMap(coordinates);
```
Comment (maxhbr) [inline modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java]: Here you make the coordinates immutable, so the second time one wants to put something (e.g. in line 253) it will fail

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/SW360Client.java [hunk 11]
```diff
@@ -26,9 +27,11 @@
 import java.util.Collections;
 
 public abstract class SW360Client {
+    private final boolean proxyUse;
     protected RestTemplate restTemplate;
 
     public SW360Client(boolean proxyUse, String proxyHost, int proxyPort) {
+        this.proxyUse = proxyUse;
```
Comment (maxhbr) [inline modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/SW360Client.java]: To be consistent with the lines below:
```suggestion
        this.proxyUse = proxyUse && proxyHost != null;
```

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/utils/RestUtils.java (hunk 8 above)
Comment (neubs-bsi) [reply modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/utils/RestUtils.java]: Will return to null, as it was before- Since this is how it is handled when called.

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java (hunk 7 above)
Comment (neubs-bsi) [reply modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/rest/resource/releases/SW360Release.java]: As I said: Returns Nullpointerexception when `(getter.apply(this));` is null

File: modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/utils/RestUtils.java (hunk 8 above)
Comment (maxhbr) [reply modules/sw360/src/main/java/org/eclipse/sw360/antenna/sw360/utils/RestUtils.java]: but if `externalIds` would be null, it would throw a NPE here?

**Comment by maxhbr** [comment]
azure is green https://dev.azure.com/EclipseSW360antenna/EclipseSW360antenna/_build/results?buildId=364
